### PR TITLE
Fix(which): match only executable files (#657)

### LINKED ISSUE #657: which() should check executability before returning a value
`which()` should only return a value if it finds a matching file in the PATH **and** that file is executable.

I'm not sure if it's possible to reliably check executability on Windows, but this should at least be done for unix. [This section](https://nodejs.org/api/fs.html#fs_file_mode_constants) of the docs suggests a way of using `fs.statSync()` to check this info.

The right way would be to also check if the current user matches the `user`, `group`, or `all` fields as well, but I think this might be tricky. It should be sufficient to just assume that we do have a match, and just do an OR of all three (ex. `userCanExecute() || groupCanExecute() || allCanExecute()`).

### PR DESCRIPTION
**Pull request by termosa**
Fixes #657.

### CODE REVIEW

File: test/which.js [hunk 1]
```diff
@@ -69,5 +70,25 @@ test('Searching with -a flag returns an array with first item equals to the regu
   t.falsy(shell.error());
   t.truthy(resultForWhich);
   t.truthy(resultForWhichA);
-  t.is(resultForWhich.toString(), resultForWhichA[0].toString());
+  t.is(resultForWhich.toString(), resultForWhichA[0]);
+});
+
+test('None executable files does not appear in the result list', t => {
+  const commandName = 'node'; // Should be an existing command
+  const pathEnv = process.env.path || process.env.Path || process.env.PATH;
```
Comment (nfischer) [inline test/which.js]: This doesn't look right. This should be `process.env.PATH`.

Only all-caps `$PATH` works on unix. On Windows, `process.env.path` is totally the same as `process.env.PATH` (and all variations of capitalization of "path").

File: test/which.js [hunk 2]
```diff
@@ -69,5 +70,25 @@ test('Searching with -a flag returns an array with first item equals to the regu
...
+  const extraPath = path.resolve(__dirname, 'resources', 'which');
+  const matchingFile = path.resolve(extraPath, commandName);
+
+  // make sure that file is exists (will throw error otherwise)
+  fs.statSync(matchingFile);
```
Comment (nfischer) [inline test/which.js]: nit: can you make this `t.truthy(fs.existsSync())`? We generally prefer assertion failures over exceptions.

File: src/which.js [hunk 3]
```diff
@@ -13,13 +13,25 @@ common.register('which', _which, {
 // set on Windows.
 var XP_DEFAULT_PATHEXT = '.com;.exe;.bat;.cmd;.vbs;.vbe;.js;.jse;.wsf;.wsh';
 
+// For earlier versions of NodeJS that doesn't have a list of constants (< v6)
```
Comment (nfischer) [inline src/which.js]: Could you add a TODO to use the node-provided constant on v6+? We'll eventually drop v4/5.

Comment (nfischer) [reply src/which.js]: Also, I'm trusting that (1) this value is respected on v4, even though the constant does not exist, and (2) the value is treated the same as `F_OK` on Windows, just as `X_OK` is documented:

> Flag indicating that the file can be executed by the calling process. This has no effect on Windows (will behave like fs.constants.F_OK).

File: test/which.js [hunk 4]
```diff
@@ -69,5 +70,25 @@ test('Searching with -a flag returns an array with first item equals to the regu
...
+
+  process.env.path = process.env.Path = process.env.PATH = extraPath + path.delimiter + pathEnv;
+  const resultForWhich = shell.which(commandName);
+  const resultForWhichA = shell.which('-a', commandName);
+  t.falsy(shell.error());
+  t.truthy(resultForWhich);
+  t.truthy(resultForWhichA);
+  t.truthy(resultForWhichA.length);
+  t.not(resultForWhich.toString(), matchingFile);
+  t.not(resultForWhichA[0], matchingFile);
```
Comment (nfischer) [inline test/which.js]: Could you rewrite this as:

```js
t.falsy(resultForWhichA.includes(matchingFile));
```

Also, I think it makes sense to skip this test on Windows, or to assert the array *does* contain the value on Windows. WDYT?

File: src/which.js (hunk 3 above)
Comment (termosa) [reply src/which.js]: True, on Windows `X_OK` and `F_OK` are equivalent. But on Windows platform file extensions are used to determine if file is executable or not.
I will not do this check for Windows platform.

File: test/which.js (hunk 4 above)
Comment (termosa) [reply test/which.js]: `includes` is not handled by Ava and it is not supported in node@5. I would also save `[0]` index to make sure that order is proper.

File: src/which.js (hunk 3 above)
Comment (nfischer) [reply src/which.js]: > I will not do this check for Windows platform.

Sounds good

**Comment by termosa** [comment]
@nfischer done

File: test/which.js [hunk 5]
```diff
@@ -69,5 +70,27 @@ test('Searching with -a flag returns an array with first item equals to the regu
...
+  const resultForWhich = shell.which(commandName);
+  const resultForWhichA = shell.which('-a', commandName);
+  t.falsy(shell.error());
+  t.truthy(resultForWhich);
+  t.truthy(resultForWhichA);
+  t.truthy(resultForWhichA.length);
+  t.not(resultForWhich.toString(), matchingFile);
+  t.is(resultForWhichA.indexOf(matchingFile), -1);
+
+  process.env.PATH = pathEnv;
```
Comment (nfischer) [inline test/which.js]: Oh sorry, I must have missed this. This should go in an `test.afterEach.always` block. Example:

https://github.com/shelljs/shelljs/blob/6d66a1af98430bc0275e2125d4cae41ccb0de645/test/exec.js#L14-L17

This shouldn't block this change. I added a comment to #834 to track this.

**Comment by nfischer** [comment]
Thanks @termosa!

**Pull request merged**